SDWebImage/Core/SDImageIOAnimatedCoder.m

```diff
@@ -208,11 +209,18 @@ + (UIImage *)createFrameAtIndex:(NSUInteger)index source:(CGImageSourceRef)sourc
         decodingOptions = [NSMutableDictionary dictionary];
     }
     CGImageRef imageRef;
-    BOOL createFullImage = thumbnailSize.width == 0 || thumbnailSize.height == 0 || pixelWidth == 0 || pixelHeight == 0 || (pixelWidth <= thumbnailSize.width && pixelHeight <= thumbnailSize.height);
+    
+    BOOL nonePixelSize = pixelWidth == 0 || pixelHeight == 0;
+    BOOL noneThumbnailSize = thumbnailSize.width == 0 || thumbnailSize.height == 0 || (pixelWidth <= thumbnailSize.width && pixelHeight <= thumbnailSize.height);
+    BOOL createFullImage = nonePixelSize || (!createThumbnail && noneThumbnailSize);
+
     if (createFullImage) {
         imageRef = CGImageSourceCreateImageAtIndex(source, index, (__bridge CFDictionaryRef)[decodingOptions copy]);
     } else {
         decodingOptions[(__bridge NSString *)kCGImageSourceCreateThumbnailWithTransform] = @(preserveAspectRatio);
+        if (noneThumbnailSize) {
+            thumbnailSize = CGSizeMake(floor(pixelWidth * 0.99), floor(pixelHeight * 0.99));
```

Comment:
Why choose `99%` width and `99%` height for this ?

Actually, if you just want something like `thumbnailRatio`, create the new option which take a double from 0.0-1.0 instead.

This bool `alwaysCreateThumbnail` does not make any scene. Since we don't know what the size of this thumbnail.

Reply:
Maybe i don't need a `thumbnailRatio`.

In fact, if possible, i always want to use `CGImageSourceCreateThumbnailAtIndex` to create images, rather than `CGImageSourceCreateImageAtIndex`. However, i found that the cache will save thumbnails and return them next time, which causes `createFullImage` always be ture.

On the other hand, if `SDImageCoderDecodeThumbnailPixelSize` is not set, or the set `SDImageCoderDecodeThumbnailPixelSize` is smaller than the pixel size, `createFullImage` will be ture.

```
BOOL createFullImage = thumbnailSize.width == 0 || thumbnailSize.height == 0 || pixelWidth == 0 || pixelHeight == 0 || (pixelWidth <= thumbnailSize.width && pixelHeight <= thumbnailSize.height);
if (createFullImage) {
    imageRef = CGImageSourceCreateImageAtIndex(source, index, (__bridge CFDictionaryRef)[decodingOptions copy]);
}
```

I think `CGImageSourceCreateThumbnailAtIndex` is a better way to create images, which can reduce memory cost.

DA LAO, is this feasible:

```
BOOL nonePixelSize = pixelWidth == 0 || pixelHeight == 0;
BOOL noneThumbnailSize = thumbnailSize.width == 0 || thumbnailSize.height == 0 || (pixelWidth <= thumbnailSize.width && pixelHeight <= thumbnailSize.height);

if (nonePixelSize) {
    imageRef = CGImageSourceCreateImageAtIndex(source, index, (__bridge CFDictionaryRef)[decodingOptions copy]);
} else {
    if (noneThumbnailSize) {
        thumbnailSize = CGSizeMake(pixelWidth, pixelHeight);
    }
    ...
    imageRef = CGImageSourceCreateThumbnailAtIndex(source, index, (__bridge CFDictionaryRef)[decodingOptions copy]);
}
```

Uh, it is unnecessary to add a `SDImageCoderOption` like `alwaysCreateThumbnail`.

Reply:
@dreampiggy

Reply:
I don't know the reason why `CGImageSourceCreateThumbnailAtIndex` on the original size of image, is anything different than call `CGImageSourceCreateImageAtIndex`

Reply:
The `thumbnail` must always be smaller (but not equal) to the full size, or it does not has anything useful meanings.

Reply:
Unless there are some known bug which cause the developer has to use `CGImageSourceCreateThumbnailAtIndex` (with full size) instead of `CGImageSourceCreateImageAtIndex` to workaround.

Reply:
uh, I test the difference in memory cost between `CGImageSourceCreateThumbnailAtIndex` and `CGImageSourceCreateImageAtIndex`.


https://user-images.githubusercontent.com/42122949/197475199-cfa287c7-068a-4f06-8462-0db64cada7b4.mov

Reply:
```
- (UIImage *)safeAnimatedImageFrameAtIndex:(NSUInteger)index {
    BOOL forceDecode = NO;
    if (@available(iOS 15, tvOS 15, *)) {
        forceDecode = YES;
    } else {
        forceDecode = NO;
    }
   ...
}
```

Below iOS 15, images are not decoded, so the memory is "busy".

Do you have any other ideas?

Reply:
@dreampiggy

Reply:
Your code snippet is because another *Bug* from Apple, not what it intended to solve.

See more in https://github.com/SDWebImage/SDWebImage/issues/3273#issuecomment-1254582874

And for the visible `RAM`, you can attach your demo. Actually, when use `Force Decode`, it allocate the RAM on the CA-render-server process, which is out-of-app process, so maybe you don't count that.

But iOS's jetsam OOM detector, will detect the physical RAM usage (not what Your App allocate), so actually whether `enable Force Decode` or `disable Force Decode`, the total RAM in physical is the same.

Reply:
You can learn more in my previous blog :https://dreampiggy.com/2019/01/18/%E4%B8%BB%E6%B5%81%E5%9B%BE%E7%89%87%E5%8A%A0%E8%BD%BD%E5%BA%93%E6%89%80%E4%BD%BF%E7%94%A8%E7%9A%84%E9%A2%84%E8%A7%A3%E7%A0%81%E7%A9%B6%E7%AB%9F%E5%B9%B2%E4%BA%86%E4%BB%80%E4%B9%88/

Reply:
Get it!

Why do I feel smoother when using `CGImageSourceCreateThumbnailAtIndex`.

This is demo:
[Demo.zip](https://github.com/SDWebImage/SDWebImage/files/9850219/Demo.zip)

Reply:
I means:

Using `CGImageSourceCreateThumbnailAtIndex` (with `pixelSize` = full size) is the same as `CGImageSourceCreateImageAtIndex` + Force Decode

So your demo should compare these two case, not `CGImageSourceCreateThumbnailAtIndex` vs `CGImageSourceCreateImageAtIndex` (without Force Decode)

Reply:
Because, CGImageSourceCreateThumbnailAtIndex does not use `vm_allocate` and lazy decoding.

CGImageSourceCreateImageAtIndex indeeded use lazy decoding (which decode image on the main queue, when Core Animation Render Server (I call it ca-render-server) process need the bitmap buffer). Decode on main queue need CPU time or MPU time

Reply:
But anway, I think the massive control is most because of Apple's bug.

We stronglly recommand to turn on the `Force Decode` for animation rendering, or it may cause huge CPU usage when display on main queue.

Reply:
> 



> I means:
> 
> Using `CGImageSourceCreateThumbnailAtIndex` (with `pixelSize` = full size) is the same as `CGImageSourceCreateImageAtIndex` + Force Decode
> 
> So your demo should compare these two case, not `CGImageSourceCreateThumbnailAtIndex` vs `CGImageSourceCreateImageAtIndex` (without Force Decode)
> 
> SD provide the force decode via `SDImageCoderHelper.defaultDecodeSolution`

But, below iOS 15, `forceDecode` seems to be fasle all the time. 
I can't find a way to use `forceDecode` on OC SDK.

Reply:
> However, I found this PR by me #2977
> 
> Which shows that, if you use non-lazy decoding `Force Decode off` for larger number of GIF, this will cause a huge OOM ratio. So at that time I just revert and use the `kCGImageSourceShouldCacheImmediately = true | kCGImageSourceShouldCache = true` instead of Force Decode solution

In this case, why not use `CGImageSourceCreateThumbnailAtIndex` to create images, or does it also cause this problem?
Maybe you can provide an option for users to choose.

Reply:
Cause problem. Some of image codec does not support thumbnail. The `CGImageSourceCreateThumbnailAtIndex` will return nil. But however, `CGImageSourceCreateImageAtIndex` not.

This is the reason why Apple ImageIO team introduce two different APIs, they do the different things.

Reply:
I means, your PR try to solve the issue `You guess is because of these two APIs call`, but I answer the question of that behavior is not because of codec, is because of *Whehther the Image is lazy-decoding*

If it's lazy decoding, it will decode on use, typically on main queue. But this can save RAM usage (allocate on demand)
If it's not lazy decoding, it will decode on global queue just when I call that API, and allocate RAM in advance

Reply:
ok, thanks.

Reply:
Let me find some better way for your feature request. Maybe a new option which means `SDImageCoderAvoidLazyDecoding`, the implementation can use either `Force Decode`, or `CGImageSourceCreateThumbnailAtIndex` API, but anyway, the returned CGImage's CGDataProvider is always not lazy access.

Reply:
👌, waiting...